File: benchmarks/binarytrees/binarytrees.js
class Node {
  constructor(left, right) { this.left = left; this.right = right; }
}
function make(depth) {
  return depth === 0 ? new Node(null, null) : new Node(make(depth - 1), make(depth - 1));
}
function check(n) {
  let total = 1;
  if (n.left) total += check(n.left);
  if (n.right) total += check(n.right);
  return total;
}
const maxDepth = 16;
console.log("stretch: " + check(make(maxDepth + 1)));
const longLived = make(maxDepth);
let sum = 0;
for (let d = 4; d <= maxDepth; d += 2) {
  const n = 1 << (maxDepth - d + 4);
  for (let i = 0; i < n; i++) sum += check(make(d));
}
console.log("sum: " + sum);
console.log("long: " + check(longLived));

File: benchmarks/binarytrees/binarytrees.rs
struct Node { left: Option<Box<Node>>, right: Option<Box<Node>> }
fn make(depth: i32) -> Box<Node> {
    if depth == 0 { Box::new(Node { left: None, right: None }) }
    else { Box::new(Node { left: Some(make(depth - 1)), right: Some(make(depth - 1)) }) }
}
fn check(n: &Node) -> i32 {
    let mut total = 1;
    if let Some(l) = &n.left { total += check(l); }
    if let Some(r) = &n.right { total += check(r); }
    total
}
fn main() {
    let max_depth = 16;
    println!("stretch: {}", check(&make(max_depth + 1)));
    let long_lived = make(max_depth);
    let mut sum: i64 = 0;
    let mut d = 4;
    while d <= max_depth {
        let n = 1 << (max_depth - d + 4);
        for _ in 0..n { sum += check(&make(d)) as i64; }
        d += 2;
    }
    println!("sum: {sum}");
    println!("long: {}", check(&long_lived));
}
